fs-index/src/index.rs

```diff
@@ -529,9 +529,8 @@ impl<Id: ResourceId> ResourceIndex<Id> {
             // If the ID has no paths, remove it from the ID to paths map
             if self.id_to_paths[&id.item].is_empty() {
                 self.id_to_paths.remove(&id.item);
+                result.removed.insert(id.item);
             }
-
-            result.removed.insert(id.item);
```

Comment:
We need some test cases for this.

As well as test cases demonstrating what happens if the caller violates assumptions:
>     /// **Note**: The caller must ensure that:
>    /// - The index is up-to-date with the file system except for the updated
>    ///   resource
>    /// - In case of a addition, the resource was not already in the index
>    /// - In case of a modification or removal, the resource was already in the
>    ///   index

I'm thinking now that we should panic when the assumptions are violated.

Reply:
> This is good, but even better would be to provide error message. User could forget to call `update_one` when resource was introduced, but make the call when it was removed.

Updated the code to gracefully return an error if this case happens